### DISCUSSION TITLE
How to efficiently calculate each element in an array or matrix?

### DISCUSSION THREAD

**Discussion by Yuxin-Lou** [josdejong/mathjs#3159] [Q&A]
First of all, this part of the code in the documentation seems out of date
![image](https://github.com/josdejong/mathjs/assets/117000057/b56ff649-2564-480f-a5d8-ceee1777e07d)
![image](https://github.com/josdejong/mathjs/assets/117000057/9d83597d-206f-4cc4-af34-1784515cb28a)
https://mathjs.org/docs/datatypes/matrices.html#arrays-and-matrices

Suppose I have an array t, and I want to use the evaluate function to perform sqrt operation on each number in it, such as evaluate("sqrt(t)"). This expression is grammatically wrong. 

So I tried the following code. I think the expression tree we built can be reused in this operation. Besides using a for loop, is there any more efficient way?
```sql 
 const t = [...Array(1000000).keys()].map(i => i + 1);
 const node = math.parse('sqrt(x)')   // returns the root Node of an expression tree
 const code = node.compile()        // returns {evaluate: function (scope) {...}}
 let startTime = performance.now();
 const a = [];
 for (var i =0;i<1000000;i++){
      a[i] = code.evaluate({x: t[i]}) 
 }
```

**Comment by josdejong** [comment]
> First of all, this part of the code in the documentation seems out of date

Thanks for letting us know, I've fixed the example.

About the performance: 

1.  Parsing once and evaluating many times is the best since parsing is the slowest part, you're doing that right.
2.  If you only use numbers, you can use the plain number implementation of all functions like `import {...} from 'mathjs/number'` instead of `from 'mathjs'`
3.  If you need more speed than that, you can see if you can implement something in plain JavaScript, I guess that something like 
    `const a = t.map(Math.sqrt)` is the fastest.
